bam_sort.c

```diff
@@ -1803,23 +1842,26 @@ typedef struct {
 // Returns 0 for success
 //        -1 for failure
 static int write_buffer(const char *fn, const char *mode, size_t l, bam1_tag *buf,
-        const sam_hdr_t *h, int n_threads, const htsFormat *fmt, char *arg_list, int no_pg)
+                        const sam_hdr_t *h, int n_threads, const htsFormat *fmt,
+                        int clear_minhash, char *arg_list, int no_pg)
 {
     size_t i;
     samFile* fp;
     fp = sam_open_format(fn, mode, fmt);
     if (fp == NULL) return -1;
-    if (!no_pg && sam_hdr_add_pg((sam_hdr_t *)h, "samtools",
-                                 "VN", samtools_version(),
+    if (!no_pg && sam_hdr_add_pg((sam_hdr_t *)h, "samtools", "VN", samtools_version(),
                                  arg_list ? "CL": NULL,
                                  arg_list ? arg_list : NULL,
                                  NULL)) {
         goto fail;
     }
-    if (sam_hdr_write(fp, (sam_hdr_t *)h) != 0) goto fail;
+    if (sam_hdr_write(fp, h) != 0) goto fail;
     if (n_threads > 1) hts_set_threads(fp, n_threads);
     for (i = 0; i < l; ++i) {
-        if (sam_write1(fp, (sam_hdr_t *)h, buf[i].bam_record) < 0) goto fail;
+        bam1_t *b = buf[i].bam_record;
+        if (clear_minhash && g_is_by_minhash && b->core.tid == -1)
+            b->core.pos = b->core.isize = -1;
```

Comment:
Why isn't this `b->core.pos = -1; b->core.isize = 0;`?

(In fact, what's this for? Simply to improve compression by flattening some noise? To undo the “unmapped read gets its mate's tid&pos” thing? It wouldn't work for the latter I think. A comment might be handy…)

Also seems like it'd be clearer for the `write_buffer` parameter to be `clear_unmapped`, the test here to be `if (clear_unmapped && b->core.tid == -1)` (so no `g_foo` used in this function) and the single non-zero argument to `write_buffer()` being adjusted accordingly — i.e., lift this `clear_minhash && g_is_by_minhash` logic out of this buffer writing function.

Reply:
You're correct, it probably should be `isize = 0`.  I was getting confused with the counting from 0 vs 1 and applying to both incorrectly.

It's there because I've modified them elsewhere to hold the two halves of the minimiser.  This is just putting them back to what they would have been, although leaving this data intact is something we wish to discuss at the file formats conference call.

As for the parameter name, it does exactly what it says on the tin and IMO `clear_unmapped` is less clear.

Reply:
> It's there because I've modified them elsewhere to hold the two halves of the minimiser.

Well, now that I know what's in the tin, this is clearer! I only skimmed the core minimiser code, so a comment here would be appreciated.

I still think lifting `clear_minhash && g_is_by_minhash` out of the function to the one significant call site would be clearer (and remove two whole instructions from the inner loop).

Reply:
I'd like to hope the compiler is capable of removing the loop invariants. :-)

Agreed though.  I've adjusted it and will push a new version as soon as I've done some trial runs.  Thanks for the review.